File: skills/tools/bilibili-analyzer/scripts/prepare.cs
#!/usr/bin/env dotnet run

using System;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

// Parse arguments (args is already available in top-level statements)
if (args.Length == 0 || args.Contains("-h") || args.Contains("--help"))
{
    PrintHelp();
    return;
}

var url = args[0];
var outputDir = GetArgValue(args, "-o", "--output") ?? ".";
var fpsStr = GetArgValue(args, "--fps") ?? "1.0";
if (!double.TryParse(fpsStr, NumberStyles.Float, CultureInfo.InvariantCulture, out var fps))
{
    Console.WriteLine($"[ERROR] Invalid fps value: {fpsStr}");
    Environment.Exit(1);
}
var videoOnly = args.Contains("--video-only");
var framesOnly = args.Contains("--frames-only");

var videoPath = Path.Combine(outputDir, "video.mp4");
var imagesDir = Path.Combine(outputDir, "images");

// Create output directory
Directory.CreateDirectory(outputDir);

Console.WriteLine(new string('=', 50));
Console.WriteLine("Bilibili Video Analyzer - Prepare Script");
Console.WriteLine(new string('=', 50));
Console.WriteLine($"URL: {url}");
Console.WriteLine($"Output: {outputDir}");
Console.WriteLine($"FPS: {fps}");
Console.WriteLine(new string('=', 50));

// Download video
if (!framesOnly)
{
    if (!await DownloadBilibiliVideoAsync(url, videoPath))
    {
        Environment.Exit(1);
    }
}

// Extract frames
if (!videoOnly)
{
    if (!File.Exists(videoPath))
    {
        Console.WriteLine($"[ERROR] Video file not found: {videoPath}");
        Environment.Exit(1);
    }

    if (!await ExtractFramesAsync(videoPath, imagesDir, fps))
    {
        Environment.Exit(1);
    }
}

Console.WriteLine();
Console.WriteLine(new string('=', 50));
Console.WriteLine("[OK] Done!");
Console.WriteLine($"  Video: {videoPath}");
Console.WriteLine($"  Images: {imagesDir}/");
Console.WriteLine(new string('=', 50));

// === Functions ===

async Task<bool> DownloadBilibiliVideoAsync(string url, string outputPath)
{
    Console.WriteLine($"[INFO] Downloading video: {url}");

    try
    {
        // Extract BV ID from URL
        var bvid = ExtractBvid(url);
        if (string.IsNullOrEmpty(bvid))
        {
            Console.WriteLine("[ERROR] Invalid Bilibili URL, cannot extract BV ID");
            return false;
        }
        Console.WriteLine($"[INFO] BV ID: {bvid}");

        using var client = CreateHttpClient();

        // Step 1: Get video info to obtain cid
        Console.WriteLine("[INFO] Fetching video info...");
        var infoUrl = $"https://api.bilibili.com/x/web-interface/view?bvid={bvid}";
        var infoJson = await client.GetStringAsync(infoUrl);
        using var infoDoc = JsonDocument.Parse(infoJson);

        var code = infoDoc.RootElement.GetProperty("code").GetInt32();
        if (code != 0)
        {
            var message = infoDoc.RootElement.GetProperty("message").GetString();
            Console.WriteLine($"[ERROR] Failed to get video info: {message}");
            return false;
        }

        var data = infoDoc.RootElement.GetProperty("data");
        var title = data.GetProperty("title").GetString();
        var cid = data.GetProperty("cid").GetInt64();
        Console.WriteLine($"[INFO] Title: {title}");
        Console.WriteLine($"[INFO] CID: {cid}");

        // Step 2: Get playback URL
        Console.WriteLine("[INFO] Fetching playback URL...");
        var playUrl = $"https://api.bilibili.com/x/player/playurl?bvid={bvid}&cid={cid}&qn=80&fnval=1";
        var playJson = await client.GetStringAsync(playUrl);
        using var playDoc = JsonDocument.Parse(playJson);

        var playCode = playDoc.RootElement.GetProperty("code").GetInt32();
        if (playCode != 0)
        {
            var message = playDoc.RootElement.GetProperty("message").GetString();
            Console.WriteLine($"[ERROR] Failed to get playback URL: {message}");
            return false;
        }

        var playData = playDoc.RootElement.GetProperty("data");
        var durl = playData.GetProperty("durl")[0];
        var videoUrl = durl.GetProperty("url").GetString();
        var size = durl.GetProperty("size").GetInt64();

        Console.WriteLine($"[INFO] Video size: {size / 1024.0 / 1024.0:F1} MB");

        // Step 3: Download video
        Console.WriteLine("[INFO] Downloading video file...");

        using var request = new HttpRequestMessage(HttpMethod.Get, videoUrl);
        request.Headers.Add("Referer", $"https://www.bilibili.com/video/{bvid}");

        using var response = await client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead);
        response.EnsureSuccessStatusCode();

        var totalBytes = response.Content.Headers.ContentLength ?? size;
        if (totalBytes <= 0) totalBytes = size > 0 ? size : 1; // Prevent division by zero

        await using var contentStream = await response.Content.ReadAsStreamAsync();
        await using var fileStream = new FileStream(outputPath, FileMode.Create, FileAccess.Write, FileShare.None, 81920, true);

        var buffer = new byte[81920]; // Larger buffer for better performance
        var totalRead = 0L;
        var lastProgress = -1;
        int bytesRead;

        while ((bytesRead = await contentStream.ReadAsync(buffer)) > 0)
        {
            await fileStream.WriteAsync(buffer.AsMemory(0, bytesRead));
            totalRead += bytesRead;

            var progress = (int)(totalRead * 100 / totalBytes);
            if (progress != lastProgress && (progress % 10 == 0 || progress == 100))
            {
                Console.WriteLine($"[INFO] Progress: {progress}%");
                lastProgress = progress;
            }
        }

        // Ensure 100% is shown
        if (lastProgress != 100)
        {
            Console.WriteLine($"[INFO] Progress: 100%");
        }

        Console.WriteLine($"[OK] Video downloaded: {outputPath}");
        return true;
    }
    catch (HttpRequestException ex)
    {
        Console.WriteLine($"[ERROR] Network error: {ex.Message}");
        return false;
    }
    catch (JsonException ex)
    {
        Console.WriteLine($"[ERROR] Failed to parse API response: {ex.Message}");
        return false;
    }
    catch (Exception ex)
    {
        Console.WriteLine($"[ERROR] Download failed: {ex.Message}");
        return false;
    }
}

string? ExtractBvid(string url)
{
    // Match BV ID from various URL formats
    // https://www.bilibili.com/video/BV1xx411c7mD
    // https://www.bilibili.com/video/BV1xx411c7mD?p=1
    // https://b23.tv/BV1xx411c7mD
    // BV1xx411c7mD
    // BV ID is exactly 12 characters: BV + 10 alphanumeric chars
    var match = Regex.Match(url, @"BV[a-zA-Z0-9]{10}");
    return match.Success ? match.Value : null;
}

HttpClient CreateHttpClient()
{
    var handler = new HttpClientHandler
    {
        AutomaticDecompression = System.Net.DecompressionMethods.GZip | System.Net.DecompressionMethods.Deflate
    };

    var client = new HttpClient(handler);
    client.DefaultRequestHeaders.Add("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36");
    client.DefaultRequestHeaders.Add("Referer", "https://www.bilibili.com");
    client.DefaultRequestHeaders.Add("Accept", "application/json, text/plain, */*");
    client.Timeout = TimeSpan.FromMinutes(30);

    return client;
}

async Task<bool> ExtractFramesAsync(string videoPath, string outputDir, double fps)
{
    Console.WriteLine($"[INFO] Extracting frames (fps={fps})");

    var ffmpeg = FindExecutable("ffmpeg", "ffmpeg.exe");
    if (ffmpeg == null)
    {
        Console.WriteLine("[ERROR] ffmpeg not found!");
        Console.WriteLine("        Windows: choco install ffmpeg / scoop install ffmpeg");
        Console.WriteLine("        macOS: brew install ffmpeg");
        Console.WriteLine("        Linux: sudo apt install ffmpeg");
        return false;
    }

    Directory.CreateDirectory(outputDir);
    var outputPattern = Path.Combine(outputDir, "frame_%04d.jpg");

    try
    {
        Console.WriteLine($"[INFO] Running ffmpeg: {ffmpeg}");

        var psi = new ProcessStartInfo
        {
            FileName = ffmpeg,
            Arguments = $"-i \"{videoPath}\" -vf \"fps={fps.ToString(CultureInfo.InvariantCulture)}\" -q:v 2 -y \"{outputPattern}\"",
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true
        };

        using var process = Process.Start(psi);
        if (process == null)
        {
            Console.WriteLine("[ERROR] Failed to start ffmpeg");
            return false;
        }

        // Read stderr asynchronously to prevent deadlock
        var stderrTask = process.StandardError.ReadToEndAsync();
        await process.WaitForExitAsync();
        var stderr = await stderrTask;

        // ffmpeg returns 0 on success, non-zero on failure
        if (process.ExitCode != 0)
        {
            // Only show last few lines of error
            var errorLines = stderr.Split('\n').TakeLast(5);
            Console.WriteLine($"[ERROR] Frame extraction failed (exit code {process.ExitCode}):");
            foreach (var line in errorLines)
            {
                if (!string.IsNullOrWhiteSpace(line))
                    Console.WriteLine($"        {line.Trim()}");
            }
            return false;
        }

        var frameCount = Directory.GetFiles(outputDir, "frame_*.jpg").Length;
        if (frameCount == 0)
        {
            Console.WriteLine("[ERROR] No frames extracted. Check if video file is valid.");
            return false;
        }

        Console.WriteLine($"[OK] Frames extracted: {frameCount} images saved to {outputDir}/");
        return true;
    }
    catch (Exception ex)
    {
        Console.WriteLine($"[ERROR] Frame extraction failed: {ex.Message}");
        return false;
    }
}

string? FindExecutable(params string[] names)
{
    var pathEnv = Environment.GetEnvironmentVariable("PATH") ?? "";
    var paths = pathEnv.Split(Path.PathSeparator);

    foreach (var name in names)
    {
        foreach (var path in paths)
        {
            try
            {
                var fullPath = Path.Combine(path, name);
                if (File.Exists(fullPath)) return fullPath;
            }
            catch { } // Ignore invalid paths
        }
    }

    // Common Windows paths
    if (OperatingSystem.IsWindows())
    {
        var commonPaths = new[]
        {
            @"C:\ffmpeg\bin",
            @"C:\Program Files\ffmpeg\bin",
            @"C:\tools\ffmpeg\bin",
            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "scoop", "shims"),
            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Microsoft", "WinGet", "Links"),
        };

        foreach (var basePath in commonPaths)
        {
            foreach (var name in names)
            {
                try
                {
                    var fullPath = Path.Combine(basePath, name);
                    if (File.Exists(fullPath)) return fullPath;
                }
                catch { }
            }
        }
    }

    // Try where/which command
    try
    {
        var cmd = OperatingSystem.IsWindows() ? "where" : "which";
        var psi = new ProcessStartInfo
        {
            FileName = cmd,
            Arguments = names[0],
            RedirectStandardOutput = true,
            UseShellExecute = false,
            CreateNoWindow = true
        };
        using var process = Process.Start(psi);
        if (process != null)
        {
            process.WaitForExit(5000); // 5 second timeout
            var output = process.StandardOutput.ReadToEnd()?.Trim();
            if (!string.IsNullOrEmpty(output))
            {
                var firstLine = output.Split('\n')[0].Trim();
                if (File.Exists(firstLine)) return firstLine;
            }
        }
    }
    catch { }

    return null;
}

string? GetArgValue(string[] args, params string[] names)
{
    for (int i = 0; i < args.Length - 1; i++)
    {
        if (names.Contains(args[i]))
        {
            return args[i + 1];
        }
    }
    return null;
}

void PrintHelp()
{
    Console.WriteLine(@"
Bilibili Video Analyzer - Prepare Script

Usage:
  dotnet run prepare.cs <url> [options]

Arguments:
  url                    Bilibili video URL (required)

Options:
  -o, --output <dir>     Output directory (default: current)
  --fps <value>          Frames per second (default: 1.0)
  --video-only           Only download video, skip frame extraction
  --frames-only          Only extract frames (requires existing video.mp4)
  -h, --help             Show this help

Examples:
  dotnet run prepare.cs ""https://www.bilibili.com/video/BV1xx411c7mD""
  dotnet run prepare.cs ""https://www.bilibili.com/video/BV1xx411c7mD"" --fps 0.5
  dotnet run prepare.cs ""https://www.bilibili.com/video/BV1xx411c7mD"" -o ./output

Requirements:
  - .NET 10 SDK
  - ffmpeg: https://ffmpeg.org/download.html
");
}

File: skills/tools/bilibili-analyzer/scripts/prepare.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Bilibili Video Downloader and Frame Extractor
Download Bilibili videos and extract frames
"""

import os
import sys
import subprocess
import argparse
import shutil

# Fix Windows console encoding
if sys.platform == "win32":
    sys.stdout.reconfigure(encoding='utf-8', errors='replace')
    sys.stderr.reconfigure(encoding='utf-8', errors='replace')


def find_yt_dlp():
    """Find yt-dlp executable or module

    Returns:
        list: Command to run yt-dlp, or None if not found
    """
    # Try direct command first
    if shutil.which("yt-dlp"):
        return ["yt-dlp"]

    # Try as Python module
    try:
        result = subprocess.run(
            [sys.executable, "-m", "yt_dlp", "--version"],
            capture_output=True,
            timeout=10
        )
        if result.returncode == 0:
            return [sys.executable, "-m", "yt_dlp"]
    except Exception:
        pass

    return None


def find_ffmpeg():
    """Find ffmpeg executable

    Returns:
        str: Path to ffmpeg, or None if not found
    """
    if shutil.which("ffmpeg"):
        return "ffmpeg"

    # Common Windows paths
    common_paths = [
        r"C:\ffmpeg\bin\ffmpeg.exe",
        r"C:\Program Files\ffmpeg\bin\ffmpeg.exe",
        r"C:\tools\ffmpeg\bin\ffmpeg.exe",
    ]
    for path in common_paths:
        if os.path.exists(path):
            return path

    return None


def download_video(url: str, output_path: str = "video.mp4") -> bool:
    """Download Bilibili video

    Args:
        url: Bilibili video URL
        output_path: Output file path

    Returns:
        Whether download succeeded
    """
    print(f"[INFO] Downloading video: {url}")

    yt_dlp_cmd = find_yt_dlp()
    if not yt_dlp_cmd:
        print("[ERROR] yt-dlp not found!")
        print("        Install with: pip install yt-dlp")
        print(f"        Current Python: {sys.executable}")
        return False

    cmd = yt_dlp_cmd + [
        "-f", "bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best",
        "-o", output_path,
        "--no-warnings",
        url
    ]

    try:
        print(f"[INFO] Running: {' '.join(cmd[:3])}...")
        result = subprocess.run(cmd, check=True)
        print(f"[OK] Video downloaded: {output_path}")
        return True
    except subprocess.CalledProcessError as e:
        print(f"[ERROR] Download failed: {e}")
        return False
    except Exception as e:
        print(f"[ERROR] Unexpected error: {e}")
        return False


def extract_frames(video_path: str, output_dir: str = "./images", fps: float = 1.0) -> bool:
    """Extract frames from video

    Args:
        video_path: Video file path
        output_dir: Output directory
        fps: Frames per second, default 1

    Returns:
        Whether extraction succeeded
    """
    print(f"[INFO] Extracting frames (fps={fps})")

    ffmpeg_cmd = find_ffmpeg()
    if not ffmpeg_cmd:
        print("[ERROR] ffmpeg not found!")
        print("        Windows: choco install ffmpeg / scoop install ffmpeg")
        print("        macOS: brew install ffmpeg")
        print("        Linux: sudo apt install ffmpeg")
        return False

    # Create output directory
    os.makedirs(output_dir, exist_ok=True)

    output_pattern = os.path.join(output_dir, "frame_%04d.jpg")

    cmd = [
        ffmpeg_cmd,
        "-i", video_path,
        "-vf", f"fps={fps}",
        "-q:v", "2",
        "-y",  # Overwrite existing files
        output_pattern
    ]

    try:
        print(f"[INFO] Running ffmpeg...")
        result = subprocess.run(cmd, check=True, capture_output=True)

        # Count generated images
        frame_count = len([f for f in os.listdir(output_dir) if f.startswith("frame_") and f.endswith(".jpg")])
        print(f"[OK] Frames extracted: {frame_count} images saved to {output_dir}/")
        return True
    except subprocess.CalledProcessError as e:
        stderr = e.stderr.decode('utf-8', errors='replace') if e.stderr else str(e)
        print(f"[ERROR] Frame extraction failed: {stderr}")
        return False
    except Exception as e:
        print(f"[ERROR] Unexpected error: {e}")
        return False


def main():
    parser = argparse.ArgumentParser(
        description="Download Bilibili video and extract frames",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  python prepare.py "https://www.bilibili.com/video/BV1xx411c7mD"
  python prepare.py "https://www.bilibili.com/video/BV1xx411c7mD" --fps 0.5
  python prepare.py "https://www.bilibili.com/video/BV1xx411c7mD" -o ./output
        """
    )

    parser.add_argument("url", help="Bilibili video URL")
    parser.add_argument("-o", "--output", default=".", help="Output directory (default: current)")
    parser.add_argument("--fps", type=float, default=1.0, help="Frames per second (default: 1)")
    parser.add_argument("--video-only", action="store_true", help="Only download video, skip frame extraction")
    parser.add_argument("--frames-only", action="store_true", help="Only extract frames (requires existing video.mp4)")

    args = parser.parse_args()

    # Set paths
    output_dir = args.output
    video_path = os.path.join(output_dir, "video.mp4")
    images_dir = os.path.join(output_dir, "images")

    # Create output directory
    os.makedirs(output_dir, exist_ok=True)

    print("=" * 50)
    print("Bilibili Video Analyzer - Prepare Script")
    print("=" * 50)
    print(f"URL: {args.url}")
    print(f"Output: {output_dir}")
    print(f"FPS: {args.fps}")
    print("=" * 50)

    # Download video
    if not args.frames_only:
        if not download_video(args.url, video_path):
            sys.exit(1)

    # Extract frames
    if not args.video_only:
        if not os.path.exists(video_path):
            print(f"[ERROR] Video file not found: {video_path}")
            sys.exit(1)

        if not extract_frames(video_path, images_dir, args.fps):
            sys.exit(1)

    print("")
    print("=" * 50)
    print("[OK] Done!")
    print(f"  Video: {video_path}")
    print(f"  Images: {images_dir}/")
    print("=" * 50)


if __name__ == "__main__":
    main()
